### ISSUE TITLE
Add Spiffe ID templating for k8s-workload-registrar

### ISSUE THREAD

**Issue by edwarnicke** [spiffe/spire#2280]
Currently the [k8s-workload-registrar](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar) allows for [creation of Spiffe IDs using CRDs](https://github.com/spiffe/spire/tree/main/support/k8s/k8s-workload-registrar#spiffe-id-custom-resources)

It would be incredibly helpful to be able to have k8s-workload-registrar to allow creating of SpiffeIDs for workloads dynamically using *templates* defined by a CRD:

```yaml
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: my-spiffe-id
  namespace: my-namespace
spec:
  spiffeId: spiffe://example.org/{{selector k8s:sa}}/{{selector k8s:pod-uid}}
  parentId: spiffe://example.org/spire/server
```

**Comment by lumjjb** [comment]
Hi @edwarnicke! @evan2645 and @dfeldman mentioned that we may have some overlap in what we're trying to do with the k8s workload registrar!

@mrsabath and I have also been planning on doing something similar to this using the k8s workload registrar, with a little bit of a different angle. We were initially thinking of writing an operator where you would define a schema, together with the source of where to get that information (and required selectors that come with them). 

We wrote a short draft document below to illustrate the idea. It would be great if there's an opportunity to collaborate!

https://github.com/IBM/trusted-service-identity/pull/95/files

**Comment by evan2645** [comment]
This would be awesome :-D

There are plans to eventually move away from k8s-workload-registrar, so I don't know if spending a huge amount of effort there will be worth the time, however I think this is a pretty lightweight mechanism that would be straightforward to implement and consume (and thus worth the effort).

I think the biggest thing to determine is where we want this configurable to live. Does it go only in CRDs? Or in labels/annotations? Or can it be configured directly on the registrar, and be used as the default ID? Or... maybe all those things? :)

Would be great to know if any of these shapes help your cause @lumjjb, and also @azdagron has expressed some thoughts here that I think he will share as soon as he gets a moment

**Comment by lumjjb** [comment]
Yea - I think this will definitely help. +1 on being able to do this in labels and annotations! That way I think it could be easier to do mutating webhook integrations for pod creation.

**Comment by azdagron** [comment]
My only thought here is that if we took something like this, it could probably be applied generally and lead to the deprecation of the current configurables we have for the pod label or annotation.

For example, the default could be something like:
```
id_path_template = "/k8s-workload-registrar/ns/{{ .Pod.ObjectMeta.Namespace }}/sa/{{ .Pod.Spec.ServiceAccountName }}"
```

And a label based one could be something like:

```
id_path_template = "/k8s-workload-registrar/{{ notempty index .Pod.ObjectMeta.Labels \"my-fancy-label\" }}"
```
where `notempty` is a template function we define that ensures that the parameter that it is passed (e.g. the label) isn't empty and fails otherwise, which allows us to not produce an ID for a pod that doesn't match our expected shape.

The above examples obviously rely on the template being passed the whole corev1.Pod object. That gives us a bunch of flexibility but maybe too much flexibility? We could also curate and filter down the information we pass to the template if we don't want to hand over the whole pod description. This would prevent the footgun of people from making IDs based on the pod status or other similar kinds of fields.

**Comment by edwarnicke** [comment]
@lumjjb Glad you like the idea! :)
 @azdagron I like where you are going with this... 

So generally, I'm pro flexibility.  But that's tempered by questions about information leaking.  So for me, the question is: is can we think of information in the Pod that *shouldn't* leak to the Spiffe ID?

What other objects might it be helpful to pass to the templating?  I can definitely see Node and Cluster (more Cluster than Node) being interesting.  Is there info from we could 'selector' on that we don't have in the .Pod (Or .Node or .Cluster)?

**Comment by lumjjb** [comment]
>What other objects might it be helpful to pass to the templating? I can definitely see Node and Cluster (more Cluster than Node) being interesting. Is there info from we could 'selector' on

It would be nice to have some information like the agent's attestor of the node it is on as well, although this seems fairly involved. 


>So generally, I'm pro flexibility. But that's tempered by questions about information leaking. So for me, the question is: is can we think of information in the Pod that shouldn't leak to the Spiffe ID?

I am too pretty flexible on this. Perhaps use of the additional information can be delegated to the operator (human or machine) could also be another option so as to not complicate things and make the registrar fairly "safe" for most needs. 

I am leaning slightly towards drawing the line of templating to the pod itself first, and see the different domain specific examples end up coming out of the use of registrar. And the registrar could slowly cherry pick things or maybe end up with some kind of plugin interface.

**Comment by edwarnicke** [comment]
> It would be nice to have some information like the agent's attestor of the node it is on as well, although this seems fairly involved.

@lumjjb Could you give some examples of templates that would use this and what they might look like?

> I am leaning slightly towards drawing the line of templating to the pod itself first, and see the different domain specific examples end up coming out of the use of registrar. And the registrar could slowly cherry pick things or maybe end up with some kind of plugin interface.

@azdagron @lumjjb Would it make sense to compromise at "Lets start with the Pod and see if other things make sense later?"

**Comment by lumjjb** [comment]
>Would it make sense to compromise at "Lets start with the Pod and see if other things make sense later?"

:+1:

>Could you give some examples of templates that would use this and what they might look like?

Based on the context [here](https://github.com/IBM/trusted-service-identity/pull/95/files#diff-9e7e52b6aa864eb9ee40b7f46c9694957452197c95c1400e137e866babc4741eR44-R51) , the way I was thinking of using it would be like 

```
id_path_template = "/k8s-workload-registrar/{{ nodeAttestor .like "aws_iid:*" "aws" || nodeAttestor .like "gcp_iit:*" "gcp" || "none" }}"
```

But yea if we just start off with pods, then this would be handled by logic somewhere else.

**Comment by faisal-memon** [comment]
We discussed this at the contributor sync. The consensus was to have a curated set of useful data from the pod (uid, name, namespace, etc) available to use to from the Spiffe ID. Has anyone started working on this?

**Comment by evan2645** [comment]
@edwarnicke I think @faisal-memon is interested in picking this up if you haven't started yet. We had a good chat with him last week about the direction.

@lumjjb we also discussed possibility of templating in the attestor name. The consensus there is that this templating feature isn't in the right spot to accomplish that effectively. Additionally, the registrar supports only k8s_psat node attestor... so we can template that in if folks want to use it, but it doesn't seem very useful at that point.

**Comment by edwarnicke** [comment]
I have not started yet.  I would be overjoyed if @faisal-memon picked this up.  I would be willing to offer whatever thoughts/advice/opinions/support/code review or general good will @faisal-memon might find of assistance to them :)

**Comment by edwarnicke** [comment]
@evan2645 @lumjjb Question: if we don't template in the attester... how would we associate a templated identity, say for a particular Pod, to that Pod?

**Comment by lumjjb** [comment]
>Question: if we don't template in the attester... how would we associate a templated identity, say for a particular Pod, to that Pod?

Wasn't in this conversation about this so my understanding may be flaws. I think the attester is referring specifically to the node attester? So templating should still be okay for workload selectors - i.e. i'm assuming that the following (below) can be done, or that selector inclusion for workload attester would be automatically included on use?


```
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: my-spiffe-id
  namespace: my-namespace
selector:
    ns: {{selector k8s:sa}}
    pod-uid: {{selector k8s:pod-uid}}
spec:
  spiffeId: spiffe://example.org/{{selector k8s:sa}}/{{selector k8s:pod-uid}}
  parentId: spiffe://example.org/spire/server
```

**Comment by edwarnicke** [comment]
@lumjjb What you have above is close to what I had in mind, but I presume you meant the ns in the selector to be sa? (or vis versa).

One question though... if a workload matches multiple selectors... how does Spire pick which one to use?

**Comment by lumjjb** [comment]
>@lumjjb What you have above is close to what I had in mind, but I presume you meant the ns in the selector to be sa? (or vis versa).

Yea, typo, let me fix that.

>One question though... if a workload matches multiple selectors... how does Spire pick which one to use?

Do you mean something like the pod containing `{{selector k8s:sa}}`, but the selector being specified as `sa:something-else`? I'm guessing in this case, would we "reject" the entry on the webhook since it doesn't quite make sense. Else i think the hardcoded selector in the `selector` field would take precedence.

**Comment by evan2645** [comment]
Hmm I think we might be talking past each other a bit here.

Why do we need to template the selectors? This issue talks about templating the resulting SPIFFE IDs - the registrar already knows what selectors it needs to target the pod.

The input to the registrar is "this specific pod is scheduled on this specific node". The output is a registration entry that is scoped to that pod uid, and parented to that node. The question mark is not how to target that pod, but what SPIFFE ID we should give that pod ... and that is where this templating feature fits in. And the data we have for the template is basically the pod spec.

There have been one or two requests in the past for more flexibility in the selectors that the registrar uses, and allowing templating there too is an interesting idea, but I think it's kind of orthogonal to the problem that this issue is trying to address?

**Comment by edwarnicke** [comment]
> Hmm I think we might be talking past each other a bit here.

I find it more likely I am misunderstanding some of how the system works :)

> The input to the registrar is "this specific pod is scheduled on this specific node". The output is a registration entry that is scoped to that pod uid, and parented to that node. 

With you so far :)

> The question mark is not how to target that pod, but what SPIFFE ID we should give that pod ... and that is where this templating feature fits in. And the data we have for the template is basically the pod spec.

This is where we get to the root of my possible misunderstanding of the system.   I think the fundamental question is: given the identity created for this Pod, how do we make sure that identity is matched *to that Pod*.  I had thought (perhaps mistakenly) that *selectors* were how you matched workloads to identities.  If I am mistaken I would be most appreciative of a description of how it really works so I can understand better :)

Please note: templating selectors isn't really my concern... that is a (possibly mistaken) means to the end of matching the identity we've made for a specific Pod to that Pod.

**Comment by evan2645** [comment]
> I find it more likely I am misunderstanding some of how the system works :)

No worries at all! Just trying to make sure we're fully aligned.

> I think the fundamental question is: given the identity created for this Pod, how do we make sure that identity is matched to that Pod. I had thought (perhaps mistakenly) that selectors were how you matched workloads to identities. If I am mistaken I would be most appreciative of a description of how it really works so I can understand better :)

You are correct, selectors are how we match identity to a pod. Registration entries have selectors defined on them, agents attest callers to find the caller's selectors, and the caller is given the identity that matches.

I think I see the confusion here

By default, the registrar (can?) compute these selectors automatically. The registrar's input includes pod uid, which is guaranteed unique to that pod, and the registrar can form a selector using this pod uid. So all the registrar is really missing from the user is a SPIFFE ID.

That said, I see now that selector can be specified as part of the SPIFFE ID CRD mode (in other modes, you do not specify selector) ... I am not sure if this is mandatory @faisal-memon? Can I use CRD mode and just let the registrar handle the selectors?

In terms of usability of CRD mode and templating and support for custom selectors ... I think that either:
* We disallow use of the templating feature with CRD mode
* We disallow use of the templating feature when CRD mode is used with custom selectors
* We apply the template to selector data as well when used in CRD mode

Otherwise we're in this weird spot with dynamic IDs and static selectors ... which perhaps raises a follow-on question which is, in this dynamic ID world, does having a CRD even make sense?

Sorry for the confusion here.

**Comment by faisal-memon** [comment]
You don't specify the selectors for auto generated entries with the CRD. It uses pod-uid+namespace+node-name to create a unique selector set. With the CRD you have the option to manually create an entry in which case you would specify the selectors.

The templating feature would only be used for auto-generated entries. The SPIFFE ID would be generated based off the template and the selectors would be auto-generated.

**Comment by edwarnicke** [comment]
@evan2645 
> By default, the registrar (can?) compute these selectors automatically. The registrar's input includes pod uid, which is guaranteed unique to that pod, and the registrar can form a selector using this pod uid. So all the registrar is really missing from the user is a SPIFFE ID.

OK... so I could decide I wanted to do a CRD like:

```yaml
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: default-spiffe-id
  namespace: my-namespace
spec:
  spiffeId: spiffe://example.org/{{selector k8s:sa}}/{{selector k8s:pod-uid}}
  parentId: spiffe://example.org/spire/server
```

and every Pod would get an identity of that form, with the selector being auto-generated by the the registrar to be something like:
```yaml
selector:
    ns: my-namespace
    node: ${node the pod withe pod-uid runs on}
    pod-uid: {{selector k8s:pod-uid}}
```

Question: Would you implicitly add the 'ns' selector above for the case where the SpiffeID has a namespace specified in its metadata?

I can easily imagine variations, where for example I might something *different* for a particular sa or set of sas or ns or set of ns

So for example, I might have:

```yaml
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: default-spiffe-id
spec:
  spiffeId: spiffe://example.org/{{selector k8s:sa}}/{{selector k8s:pod-uid}}
  parentId: spiffe://example.org/spire/server
```

which would implicitly generate a selector in the registrar like:

```yaml
selector:
    node: ${node the pod withe pod-uid runs on}
    pod-uid: {{selector k8s:pod-uid}}
```

But what if I want Pods in sa "my-sa" to still be templetized?

Could I *also* have a SpiffID like:

```yaml
apiVersion: spiffeid.spiffe.io/v1beta1
kind: SpiffeID
metadata:
  name: my-sa-spiffe-id
selector:
    sa: my-sa
spec:
  spiffeId: spiffe://example.org/special/{{selector k8s:sa}}/{{selector k8s:pod-uid}}
  parentId: spiffe://example.org/spire/server
```

If I did, what would the registrar do for the selector?  It feels like it would simply combine the 'static' `sa: my-sa` selector with the implicit one.  Which brings me back to my question: how do we decide which selector wins if they overlap?

**Comment by faisal-memon** [comment]
@edwarnicke What im thinking is simpler. The k8s-workload-registrar would have a config option like:

`spiffeid_template = /ns/{{namespace}}/sa/{{serviceaccount}}`

Then pods with a label/annotation of `spiffe.io/spiffeid: true` would have entries auto-generated with the SPIFFE ID following the specified format. The selectors would be auto generated using the same pod-uid+namespace+node-name set. The pod-uid is the main one that guarantees uniqueness.

**Comment by edwarnicke** [comment]
@faisal-memon So are you thinking one template per Cluster?  If not, how do I know which template to apply when?

**Comment by edwarnicke** [comment]
@faisal-memon I have use cases where I *definitely* want templating and *definitely don't* want one template per cluster.

I'm perfectly delighted to have what you describe as the simple out the gate path, but I *really* need a way to deal with exceptions.

@evan2645 How does Spire deal *now* with a workload that matches multiple selectors?

**Comment by faisal-memon** [comment]
@edwarnicke Maybe a good time to take a step back and understand your use case better. Why do you need more than one template per cluster?

**Comment by edwarnicke** [comment]
@faisal-memon It fundamentally comes down to domains of administrative control.

Here's *one* example:

I am federating with an administratively distinct entity (think another company).  They have strict standards about the format of spiffeIds they want, because they are building policy for them.  I have existing standards for the format of spiffeIds I'm using already, that *I* can't change.

I can also tell you from working in a huge company... it doesn't even have to be a big company... different departments within a large company will *also* have definitive opinions that will diverge.

Your proposal is a 'one template per cluster' standard... and while I *love* that as the default path... what cluster I am running in is a completely non-sequitur in the generic as to what the desired template for my spiffeId should be.

The real question is:
- Does choice of cluster imply template for SpiffeID?  I would maintain it doesn't.  Choice of cluster is a non-sequitur to templale of SpiffeID in the generic.

**Comment by edwarnicke** [comment]
@faisal-memon Another way to think of this is in terms of "What Scope has to agree?"

If my template is *only* cluster scope... *all* users of the cluster must agree on that template.

I don't see any reason why 'cluster' is the natural Scope of agreement on template.

**Comment by lumjjb** [comment]
To kind of follow the kubernetes demarkation, would namespace and cluster wide be an appropriate scope? I can see the potential of there being a cluster wide resource to define templating for the entire cluster as well as namespaced resource that defines that for the namespace. If both co-exists it would probably still be okay as well since each pod would get both SVIDs.

This brings up a question of the threat model though, either different departments/partition need to trust that the other is non malicious, or there would need to have the controller do some policy control (or perhaps do a controller installation per namespace, and each registrar has its own SVID for separate policy authentication via something like https://github.com/spiffe/spire/issues/1975).

**Comment by edwarnicke** [comment]
> To kind of follow the kubernetes demarkation, would namespace and cluster wide be an appropriate scope?
I could easily see namespace being *one* of the demarkations... but I could also easily see service account being the demark as well.

**Comment by edwarnicke** [comment]
@lumjjb 
>

**Comment by edwarnicke** [comment]
@lumjjb 
> To kind of follow the kubernetes demarkation, would namespace and cluster wide be an appropriate scope?

namespace is definitely *one* of the demarks you might choose... absolutely.  But service account is another one, and selector by labels is an appropriate third option.  Basically... k8s thinks in terms of selectors... and so identity template probably wants to enable the likely ones.

>  If both co-exists it would probably still be okay as well since each pod would get both SVIDs.
Which SVID does the Pod use then to communicate with things?  I had the (perhaps mistaken) notion that a Pod had zero or one identities... I was unaware that it had the potential for zero to n identities, and I'm not sure that most workloads would be smart enough to cope with that.

> This brings up a question of the threat model though, either different departments/partition need to trust that the other is non malicious, or there would need to have the controller do some policy control (or perhaps do a controller installation per namespace, and each registrar has its own SVID for separate policy authentication via something like #1975).

Or have the cluster administrator configure the templates (similar to what I'd do if I wanted a service identity associated identity)

What I'm having a bit of confusion on though is, given that Spire has this really really nice selector mechanism... why not use that in some way to reduce the introduction of new semantics unless absolutely needed.

**Comment by faisal-memon** [comment]
Sorry for the delayed reply, I was thinking this over. I like the proposal from @lumjjb. Have cluster and namespace scoped templates. Im thinking to have a cluster wide default and the ability to override and use a specific template for a namespace.

@edwarnicke doesn't seem like that will fully meet your needs. Maybe we can start with this and see how it goes?

**Comment by edwarnicke** [comment]
@lumjjb Add service account scoped and I can get onboard :)

**Comment by edwarnicke** [comment]
I am still trying to understand though how a single workload gets issued multiple identities and what we expect a workload to do if that happens...

**Comment by lumjjb** [comment]
@edwarnicke Hmm, how would you define it for service account scopes? Would it be part of service account object annotations? or something along those lines?

**Comment by edwarnicke** [comment]
@lumjjb Meaning: I can apply a different template to different service accounts.

**Comment by lumjjb** [comment]
Gotcha!

@mrsabath is working on implementing a customizable template prototype
which may be extensible to do that. I think he can share a bit about that
soon.

**Comment by edwarnicke** [comment]
@lumjjb Awesome!  I appreciate your patience with my requests :)

**Comment by lumjjb** [comment]
Excited to see what we end up with. Been pretty good discussions so far!

**Comment by mrsabath** [comment]
Hi @edwarnicke and @faisal-memon. This has been a productive discussion! 
@lumjjb and I created a quick prototype that is trying to address some of the aspects addressed here. 

This extension to the existing workload registrar controllers provides a function to template the format of the SVIDs. It basically takes a template (a configuration) file: `identity-schema.yaml`, and formats the SVID based on the fields provided in this template. Values for the fields can be retrieved from various components, e.g. workload attestor, node attestor (not implemented), or from arbitrary configMaps. In addition, this code also modifies the relevant selector values. This allows flexibility for defining the final format of the SVID.

In the provided example, there is only one format per cluster instance, but this can be also implemented per namespace. This code is intended to run in `crd-mode` and currently does not modify the format of the SVID that is provided via pod label or annotation. It simply changes the default value, from: 
```
"spiffe://$TrustDomain/ns/$namespace/sa/$serviceAccount"
to, e.g.: 
"spiffe://$TrustDomain/$serviceProvider/$region/$namespace/$serviceAccount/$pod-name"
```
This template format allows us to use various dynamic policies for managing identities.
Our code is available here: 
https://github.com/mrsabath/spire/compare/main...mrsabath:identity-schema

We would be really happy to discuss this further and find a solution that is general for broad use-case usage.

**Comment by edwarnicke** [comment]
@mrsabath This is fantastic news!  Do we have a PR yet?  I'd love to look at the code/kick the tires :)

I am curious, what is the impediment to also doing this scoped per Service Account?

**Comment by mrsabath** [comment]
@edwarnicke there is no formal PR yet, as we are still developing the prototype, but the link above should give you a preview of what we are thinking. In fact, your feedback would help us crystalize the actual next steps. 

Regarding the Service Account, it's just an example of what could be the various elements for defining the identity.

**Comment by edwarnicke** [comment]
@mrsabath I think it's important to distinguish 'data you can fill into the identity template' from 'scope of template applicability'

What I was asking for is Service Account as a 'scope of applicability.  Meaning I can have a distinct template for my Service Account.

**Comment by mrsabath** [comment]
@edwarnicke there should not be a problem to scope the templates across Service Accounts. We started with the simplest example, but I see your point, and that's a great suggestion that provides new use cases.
 
Different question, do you see a value in introducing a pod(deployment, etc) level flag that enables which format template to use? Similar to what pod label/annotations are doing now?

**Comment by edwarnicke** [comment]
@mrsabath I'm all about starting easy and building from there :)

> Different question, do you see a value in introducing a pod(deployment, etc) level flag that enables which format template to use? Similar to what pod label/annotations are doing now?

Do you mean this to be the way to select among multiple templates that would otherwise match the Pod?

**Comment by azdagron** [comment]
Solved by #2417.
